### PR TITLE
Unseal support

### PR DESCRIPTION
**Pull request by magik6k**

### CODE REVIEW

**Comment by mishmosh** [comment]
Hi @magik6k do you know when you expect this to land? Hoping to use it in https://github.com/filecoin-project/go-fil-markets/issues/222.

File: ffiwrapper/partialfile.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+type partialFile struct {
+	maxPiece abi.UnpaddedPieceSize
+
+	path      string
+	allocated rlepluslazy.RLE
+
+	file *os.File
+}
+
+func writeTrailer(psz int64, w *os.File, r rlepluslazy.RunIterator) error {
```
Comment (whyrusleeping) [inline ffiwrapper/partialfile.go:35]: what is `psz`? can i get a comment?

File: ffiwrapper/sealer_cgo.go [hunk 2]
```diff
@@ -114,82 +136,183 @@ func (cf closerFunc) Close() error {
...
+		outpipe, err := ioutil.TempFile(os.TempDir(), "sector-storage-unseal-")
+		if err != nil {
+			return xerrors.Errorf("creating temp pipe file: %w", err)
+		}
+		var outpath string
+		var perr error
+		outWait := make(chan struct{})
+
+		{
+			outpath = outpipe.Name()
```
Comment (whyrusleeping) [inline ffiwrapper/sealer_cgo.go:220]: what

Comment (magik6k) [reply ffiwrapper/sealer_cgo.go:220]: IIRC @laser was working on making UnsealRange accept an FD, which would make this much , much nicer. Not sure what the progress on this is

File: fr32/fr32.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+			at = next
+		}
+
+		out[outOff+31] |= at << 6
+
+		for i := 32; i < 64; i++ {
+			next := in[i+inOffNext]
+
+			out[outOff+i] = at >> 2
+			out[outOff+i] |= next << 6
```
Comment (magik6k) [inline fr32/fr32.go:111]: This is apparently much faster than
```
out[outOff+i] = at >> 2 | next << 6
```

File: ffiwrapper/sealer_cgo.go [hunk 4]
```diff
@@ -232,6 +385,17 @@ func (sb *Sealer) SealPreCommit1(ctx context.Context, sector abi.SectorID, ticke
 		return nil, xerrors.Errorf("aggregated piece sizes don't match sector size: %d != %d (%d)", sum, ussize, int64(ussize-sum))
 	}
 
+	//staged := filepath.Join(paths.Cache, "staged")
+
+	/*if err := sb.rewriteAsPadded(paths.Unsealed, staged); err != nil {
```
Comment (whyrusleeping) [inline ffiwrapper/sealer_cgo.go:390]: this commented out stuff need to stay?

File: fr32/fr32.go (hunk 3 above)
Comment (whyrusleeping) [reply fr32/fr32.go:130]: :thinking:

**Pull request merged**